// Copyright 2024 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.


syntax = "proto3";
package datacommons;

// This proto contains specification of a statsVar Observation spec.
// It contains the stat_type, measurement property, measurement denominator,
// measurement qualifier, scaling factor, and anme of the spec if shown as a
// leaf node on the PV tree
message ObsProp {
  optional string mprop = 2;
  optional string mqual = 3;
  optional string mdenom = 4;
  reserved 1, 6, 7;
}

// This proto contains the specification of a PopObs instance. It contains
// the intended population type, constraining, property and hidden
// property-value pairs and a list of ObsProps
message PopObsSpec {
  // A property-value pair
  message PV {
    optional string prop = 1;
    optional string val = 2;
  }

  // Population type.
  optional string pop_type = 1;
  // Constraining properties of StatisticalPopulation.
  repeated string cprop = 4;
  // A list of depending property value pairs that a client does not indicate
  // but needs to be added.
  // For example, property "income" needs additional pv of "age=Years15Onwards"
  repeated PV dpv = 5;
  // Verticals of this spec.
  repeated string vertical = 7;
  repeated ObsProp obs_props = 8;

  // Set this if SVs/SVGs satisfying this spec should only appear under the
  // listed verticals. For example, this is used to ensure prison-related
  // categories only show up under crime and not under Demographics->Gender.
  //
  // Be careful in using this. If two co-occurring constraint props (say, gender
  // and prisonSentence) are under different verticals and both have this flag
  // set, then that combination of constraint props will not appear in the Tree
  // Widget at all!
  //
  // This is also meaningful with 0-P or 1-P specs only.
  optional bool show_under_listed_verticals_only = 9;
  reserved 6;
}

// All published Pop/Obs spec. Stored in a textproto.
message PopObsSpecList {
  repeated PopObsSpec spec = 1;
}
